_posts/2025-11-12-adaptive-refresh-for-resilient-segment-replication.md

```diff
@@ -0,0 +1,212 @@
+
+The single-checkpoint method forces each replica to absorb all accumulated changes in one step. It's simple and correct, but when the gap between checkpoints absorbed by the replica is large, you pay the whole cost at once. This cliff-like behavior is exactly what we set out to smooth out.
+
+
+## Bite-sized commits and adaptive refresh
+
+With *“bite-sized commits”*, we rethink how checkpoints are constructed and consumed. Instead of bundling only the newest commit, the writer now keeps a small rolling history of commits and publishes them together as a single checkpoint. On the search side, replicas receive a set of commits instead of a single one. They can step through those commits incrementally, downloading manageable deltas and refreshing in small, predictable cycles. 
+
+Keen readers would note however, that Lucene’s `refreshIfNeeded` API only refreshed on the latest commit by default. We [contributed changes](https://github.com/apache/lucene/pull/14443) to Lucene, that allow `SearcherManager` (and related classes) to intelligently select the candidate commit for refresh. Consumer implementations can add logic to select the right commit, like calculating the bytes difference between current searcher commit and available commits, and selecting the newest commit that fits in the available searcher memory. Replicas can thus *adaptively refresh* on the bite sized commits. If all commits are small enough, they can directly jump to the latest one. At the other extreme, they can incrementally step through all provided commits. These changes are available as part of Lucene 10.3.
+
+
+### Setting up indexers for bite sized commits
+
+On the indexer side, Lucene's IndexWriter continues to operate as before, creating commits on the indexer machine, though you may want to consider increasing commit frequency to keep each commit as a smaller differential from the previous one. For example, instead of creating a commit every minute, you can consider creating one every 15 seconds, giving you 4 incremental hops per minute window that a replica can chose from. Note that this is one of many possible approaches. Another, better way is to accurately trigger commits whenever segment turnover exceeds a configured threshold size.
+
+`IndexWriters` are configured with an `IndexDeletionPolicy` to clean up older commits that are no longer needed. The default is `KeepOnlyLastCommitDeletionPolicy`, which only retains the most recent commit. Since we want to hold multiple commits in our checkpoint, we must use a different index deletion policy. Lucene offers `KeepLastNCommitsDeletionPolicy` to keep last N commits. You can also write your own deletion policy based on specific business requirements, like keeping all commits within a time window.
```

Comment:
```suggestion
`IndexWriters` are configured with an `IndexDeletionPolicy` to clean up older commits that are no longer needed. The default is `KeepOnlyLastCommitDeletionPolicy`, which only retains the most recent commit. Since we want to hold multiple commits in our checkpoint, we must use a different index deletion policy. Lucene offers `KeepLastNCommitsDeletionPolicy` to keep the last N commits. You can also write your own deletion policy based on specific business requirements, like keeping all commits within a specified time window.
```